test/Microsoft.AspNetCore.ResponseCompression.Tests/ResponseCompressionMiddlewareTest.cs

```diff
@@ -0,0 +1,298 @@
+
+            CheckResponseCompressed(response, expectCompressed: expectCompressed, expectedBodyLength: expectedBodyLength);
+        }
+
+        [Fact]
+        public async Task Request_UnauthorizedMimeType()
+        {
+            var response = await InvokeMiddleware(100, requestAcceptEncoding: "gzip", responseType: "text/html");
+
+            CheckResponseCompressed(response, expectCompressed: false, expectedBodyLength: 100);
+        }
+
+        [Fact]
+        public async Task Request_ResponseWithContentRange()
+        {
+            var response = await InvokeMiddleware(50, requestAcceptEncoding: "gzip", responseType: TextPlain, addResponseAction: (r) =>
+            {
+                r.Headers[HeaderNames.ContentRange] = "1-2/*";
+            });
+
+            CheckResponseCompressed(response, expectCompressed: false, expectedBodyLength: 50);
+        }
+
+        [Fact]
+        public async Task Request_ResponseWithContentEncodingAlreadySet()
+        {
+            var otherContentEncoding = "something";
+
+            var response = await InvokeMiddleware(50, requestAcceptEncoding: "gzip", responseType: TextPlain, addResponseAction: (r) =>
+            {
+                r.Headers[HeaderNames.ContentEncoding] = otherContentEncoding;
+            });
+
+            Assert.Equal("MD5", response.Headers[HeaderNames.ContentMD5]);
+            Assert.Equal(otherContentEncoding, response.Headers[HeaderNames.ContentEncoding]);
+            Assert.Null(response.ContentLength);
+            Assert.Equal(50, response.Body.Length);
+        }
+
+        [Theory]
+        [InlineData(false, 100)]
+        [InlineData(true, 24)]
+        public async Task Request_Https(bool enableHttps, int expectedLength)
+        {
+            var options = new ResponseCompressionOptions()
+            {
+                MimeTypes = new string[] { TextPlain },
+                Providers = new IResponseCompressionProvider[]
+                {
+                    new GzipResponseCompressionProvider(CompressionLevel.Optimal)
+                },
+                EnableHttps = enableHttps
+            };
+
+            var middleware = new ResponseCompressionMiddleware(async context =>
+            {
+                context.Response.ContentType = TextPlain;
+                await context.Response.WriteAsync(new string('a', 100));
+            }, Options.Create(options));
+
+            var httpContext = new DefaultHttpContext();
+            httpContext.Request.Headers[HeaderNames.AcceptEncoding] = "gzip";
+            httpContext.Request.IsHttps = true;
+
+            httpContext.Response.Body = new MemoryStream();
+
+            await middleware.Invoke(httpContext);
+
+            Assert.Equal(expectedLength, httpContext.Response.Body.Length);
+        }
+
+        private async Task<HttpResponse> InvokeMiddleware(int uncompressedBodyLength, string requestAcceptEncoding, string responseType, Action<HttpResponse> addResponseAction = null)
```

Comment:
I have already tried to use `TestServer`, but I cannot neighter check the Content-Encoding header value in `HttpResponseMessage` (it is empty) nor the content length (the decompression is automatically done). Or I miss something?